### PR TITLE
soc: nrf: move z_platform_init to a common directory and prevent stack usage before SystemInit is called

### LINKED ISSUE #36351: nRF: we do not always guarantee that SystemInit is inlined
**Describe the bug**
In nRF platforms, SystemInit() is invoked by z_platform_init, during early reset cycle. Normally SRAM should not be accessed before SystemInit has returned, so we should be ensuring that no push register instructions are executed when branching to SystemInit(). Currently we do not ensure this under all optimization regimes. This needs to be fixed.

**Impact**
Unclear. It is however specified that SRAM access needs to be avoided before system is initialized, so we better comply with the standard.

### PR DESCRIPTION
**Pull request by ioannisg**
Use a common implementation for z_platform_init
in nRF SoC family. The implementations are
identical, so no need to maintain separate
identical versions of them.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Fixes #36351

### CODE REVIEW

File: soc/arm/nordic_nrf/common/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_include_directories(.)
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/common/CMakeLists.txt:4]: probably not needed

File: soc/arm/nordic_nrf/common/soc_nrf_common.S [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+GTEXT(z_platform_init)
+
+SECTION_FUNC(TEXT, z_platform_init)
+
+#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
+	ldr r0, =SystemInit
+	bx r0
+#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	b SystemInit
+#endif
```
Comment (anangl) [inline soc/arm/nordic_nrf/common/soc_nrf_common.S:26]: An `#else` here would help prevent `z_platform_init` from ending up without any implementation in an unsupported configuration.

Comment (ioannisg) [reply soc/arm/nordic_nrf/common/soc_nrf_common.S:26]: I know... I also thought about this, right. But currently nRF families only have cortex-m, and cortex-m only has these 3 ARM architecture variants, so I was a bit reluctant to add and #error "unsupported arch".  :) 
But I can add it now that you also suggested this. no big deal

**Review by anangl — APPROVED** [review]

**Review by oyvindronningstad — APPROVED** [review]
Could this achieved with __attribute__((naked))?

If not, I'd prefer a naked C function, but it's up to you.

In any case, add a comment in the code about why this is not done as a regular function.

**Comment by ioannisg** [comment]
> Could this achieved with **attribute**((naked))?
> 
> If not, I'd prefer a naked C function, but it's up to you.
> 
> In any case, add a comment in the code about why this is not done as a regular function.

Thanks for the review @oyvindronningstad , I'll add a comment in the the code about the rationale behind choosing ASM.

This is not user code - so, honestly, it is OK to use ASM if this is solving our problem and making things very clear to read. C function attributions are more complicated than basic assembly, imho.

**Comment by ioannisg** [comment]
Added the comment in the code. And btw, the rationale for doing this in ASM is also present in the commit message.

**Review by oyvindronningstad — APPROVED** [review]

**Pull request merged**